File: agent/src/main/python/hms_agent/Runner.py
#!/usr/bin/env python

'''
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
'''

import threading
from daemon import daemonRunner
from package import packageRunner
from shell import shellRunner

class Runner(threading.Thread):
    __instance = None
    lock = None
    def __init__(self):
        if Runner.__instance is None:
          Runner.lock = threading.RLock()
          Runner.__instance = self
    
    def run(self, data):
        Runner.lock.acquire()
        try:
            if data['actionType']=='info':
                ph = packageRunner()
                result = ph.info(data['packages'])
            elif data['actionType']=='install':
                ph = packageRunner()
                if 'dry-run' in data:
                    opt = data['dry-run']
                else:
                    opt = 'false'
                result = ph.install(data['packages'], opt)
            elif data['actionType']=='remove':
                ph = packageRunner()
                if 'dry-run' in data:
                    opt = data['dry-run']
                else:
                    opt = 'false'
                result = ph.remove(data['packages'], opt)
            elif data['actionType']=='status':
                dh = daemonRunner()
                result = dh.status(data['daemonName'])
            elif data['actionType']=='start':
                dh = daemonRunner()
                result = dh.start(data['daemonName'])
            elif data['actionType']=='stop':
                dh = daemonRunner()
                result = dh.stop(data['daemonName'])
            elif data['actionType']=='run' or data['@action']=='org.apache.hms.common.entity.action.ScriptAction':
                she = shellRunner()
                script = []
                script.append(data['script'])
                if "parameters" in data:
                    for parameter in data['parameters']:
                        script.append(parameter)
                result = she.run(script)
            return result
        finally:
            Runner.lock.release()

File: contrib/fast-hdfs-resource/src/main/java/org/apache/ambari/fast_hdfs_resource/Runner.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ambari.fast_hdfs_resource;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.fs.FileSystem;

import com.google.gson.Gson;

public class Runner {
  public static void main(String[] args)
      throws IOException, URISyntaxException {
    // 1 - Check arguments
    if (args.length != 1) {
      System.err.println("Incorrect number of arguments. Please provide:\n"
          + "1) Path to json file\n"
          + "Exiting...");
      System.exit(1);
    }

    // 2 - Check if json-file exists
    final String jsonFilePath = args[0];
    File file = new File(jsonFilePath);

    if (!file.isFile()) {
      System.err
          .println("File " + jsonFilePath + " doesn't exist.\nExiting...");
      System.exit(1);
    }

    Gson gson = new Gson();
    Resource[] resources = null;
    FileSystem dfs = null;

    try {
      Configuration conf = new Configuration();
      dfs = FileSystem.get(conf);

      // 3 - Load data from JSON
      resources = (Resource[]) gson.fromJson(new FileReader(jsonFilePath),
          Resource[].class);

      // 4 - Connect to HDFS
      System.out.println("Using filesystem uri: " + FileSystem.getDefaultUri(conf).toString());
      dfs.initialize(FileSystem.getDefaultUri(conf), conf);
      
      for (Resource resource : resources) {
        System.out.println("Creating: " + resource);

        Resource.checkResourceParameters(resource, dfs);

        Path pathHadoop = new Path(resource.getTarget());
        if (!resource.isManageIfExists() && dfs.exists(pathHadoop)) {
          System.out.println("Skipping the operation for not managed DFS directory " + resource.getTarget() +
                             " since immutable_paths contains it.");
          continue;
        }

        if (resource.getAction().equals("create")) {
          // 5 - Create
          Resource.createResource(resource, dfs, pathHadoop);
          Resource.setMode(resource, dfs, pathHadoop);
          Resource.setOwner(resource, dfs, pathHadoop);
        } else if (resource.getAction().equals("delete")) {
          // 6 - Delete
          dfs.delete(pathHadoop, true);
        }
      }
    } 
    catch(Exception e) {
       System.out.println("Exception occurred, Reason: " + e.getMessage());
       e.printStackTrace();
    }
    finally {
      dfs.close();
    }

    System.out.println("All resources created.");
  }

}
